### DISCUSSION TITLE
Wifi status led usage

### DISCUSSION THREAD

**Discussion by schortl** [arendst/Tasmota#13356] [Q&A] [answered]
### PROBLEM DESCRIPTION
Wifi status led is not properly assigned (my devices are SHP5 and SP111):
In case of two LEDs it is described that if more than one LED component is defined, Led1/Led1i will act as the Wi-Fi/MQTT status LED and the next defined LED (e.g., Led2/Led2i) will act as the LED that indicates the power state of the relay(s). 
It seems that since release 8.5.1 Led1/Led1i is not used as Wi-Fi/MQTT status LED but is ignored and not controlled (in release 7.2.0 it operates as described).

In release 8.5.1  it seems that no  Wi-Fi/MQTT status indication is possible (Wi-Fi and MQTT connection indication IS enabled via setoption31 = 0 !)  and in release  9.3.0 and current release 9.5.0 Led2/Led2i instead of Led1/Led1i acts as Wi-Fi/MQTT status LED AND in addition to the power state indication . Led1/Led1i  is ignored (without function).

regards
shortl

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [ ] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [ ] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [ ] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [ ] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): SHP5 and SP111
- [  ] Tasmota binary firmware version number used: 8.51, 9.3.0, 9.5.0
  - [x] Pre-compiled
  - [ ] Self-compiled
- [x] Flashing tools used:  wired flash (tasmotizer)
- [x] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:
23:18:15 RSL: RESULT = {"NAME":"Nous A1","GPIO":[57,0,56,0,132,134,0,0,131,17,0,21,0],"FLAG":0,"BASE":45}
23:18:15 RSL: RESULT = {"Module":{"0":"Nous A1"}}
23:18:16 RSL: RESULT = {"GPIO0":{"57":"Led2i"},"GPIO1":{"0":"None"},"GPIO2":{"56":"Led1i"},"GPIO3":{"0":"None"},"GPIO4":{"132":"HLWBL CF1"},"GPIO5":{"134":"BL0937 CF"},"GPIO9":{"0":"None"},"GPIO10":{"0":"None"},"GPIO12":{"131":"HLWBL SELi"},"GPIO13":{"17":"Button1"},"GPIO14":{"0":"None"},"GPIO15":{"21":"Relay1"},"GPIO16":{"0":"None"}}

```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [x] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:
23:19:18 RSL: STATUS = {"Status":{"Module":0,"DeviceName":"Nous switch 2","FriendlyName":["Nous 2 230V"],"Topic":"tasmota_64BFAC","ButtonTopic":"0","Power":0,"PowerOnState":0,"LedState":1,"LedMask":"FFFF","SaveData":0,"SaveState":1,"SwitchTopic":"0","SwitchMode":[0,0,0,0,0,0,0,0],"ButtonRetain":0,"SwitchRetain":0,"SensorRetain":0,"PowerRetain":0}}
23:19:18 RSL: STATUS1 = {"StatusPRM":{"Baudrate":115200,"SerialConfig":"8N1","GroupTopic":"tasmotas","OtaUrl":"http://ota.tasmota.com/tasmota/release/tasmota.bin.gz","RestartReason":"Software/System restart","Uptime":"0T00:28:09","StartupUTC":"2021-10-13T20:51:09","Sleep":50,"CfgHolder":4617,"BootCount":26,"BCResetTime":"2021-10-11T09:49:19","SaveCount":90,"SaveAddress":"F4000"}}
23:19:18 RSL: STATUS2 = {"StatusFWR":{"Version":"8.5.1(tasmota)","BuildDateTime":"2020.10.02 10:15:28","Boot":31,"Core":"2_7_4_1","SDK":"2.2.2-dev(38a443e)","CpuFrequency":80,"Hardware":"ESP8285","CR":"396/699"}}
23:19:18 RSL: STATUS3 = {"StatusLOG":{"SerialLog":2,"WebLog":2,"MqttLog":0,"SysLog":0,"LogHost":"","LogPort":514,"SSId":["Simpsons",""],"TelePeriod":300,"Resolution":"558180C0","SetOption":["00008001","2805C8000100060000005A00000000000000","00000000","00006000","00000000"]}}
23:19:18 RSL: STATUS4 = {"StatusMEM":{"ProgramSize":597,"Free":404,"Heap":26,"ProgramFlashSize":1024,"FlashSize":1024,"FlashChipId":"144051","FlashFrequency":40,"FlashMode":3,"Features":["00000407","8FDAE797","04368001","000000CD","010013C0","C000F981","00004004","00000000"],"Drivers":"1,2,3,4,5,6,7,8,9,10,12,16,18,19,20,21,22,24,26,27,29,30,35,37","Sensors":"1,2,3,4,5,6"}}
23:19:18 RSL: STATUS5 = {"StatusNET":{"Hostname":"tasmota_64BFAC-8108","IPAddress":"192.168.1.142","Gateway":"192.168.1.1","Subnetmask":"255.255.255.0","DNSServer":"192.168.1.1","Mac":"70:03:9F:64:BF:AC","Webserver":2,"WifiConfig":4,"WifiPower":17.0}}
23:19:18 RSL: STATUS7 = {"StatusTIM":{"UTC":"2021-10-13T21:19:18","Local":"2021-10-13T23:19:18","StartDST":"2021-03-28T02:00:00","EndDST":"2021-10-31T03:00:00","Timezone":99,"Sunrise":"07:29","Sunset":"18:29"}}
23:19:18 RSL: STATUS9 = {"StatusPTH":{"PowerDelta":[0,0,0],"PowerLow":0,"PowerHigh":0,"VoltageLow":0,"VoltageHigh":0,"CurrentLow":0,"CurrentHigh":0}}
23:19:18 RSL: STATUS10 = {"StatusSNS":{"Time":"2021-10-13T23:19:18","ENERGY":{"TotalStartTime":"2021-10-11T09:49:19","Total":0.000,"Yesterday":0.000,"Today":0.000,"Power":0,"ApparentPower":0,"ReactivePower":0,"Factor":0.00,"Voltage":0,"Current":0.000}}}
23:19:18 RSL: STATUS11 = {"StatusSTS":{"Time":"2021-10-13T23:19:18","Uptime":"0T00:28:09","UptimeSec":1689,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"Sxxxxs","BSSId":"3C:39:12:09:19:F4","Channel":6,"RSSI":96,"Signal":-52,"LinkCount":1,"Downtime":"0T00:00:03"}}}

```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:


```

### TO REPRODUCE
_Steps to reproduce the behavior:_
flash and configure device. Problem arises when no WiFi signal is received. 

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._
 if more than one LED component is defined, Led1/Led1i will act as the Wi-Fi/MQTT status LED and the next defined LED (e.g., Led2/Led2i) will act as the LED that indicates the power state of the relay(s). 
It seems that since release 8.5.1 Led1/Led1i is not used as Wi-Fi/MQTT status LED but is ignored and not controlled (in release 7.2.0 it operates as described).

In release 8.5.1  it seems that no  Wi-Fi/MQTT status indication is possible (Wi-Fi and MQTT connection indication IS enabled via setoption31 = 0 !)  and in release  9.3.0 and current release 9.5.0 Led2/Led2i instead of Led1/Led1i acts as Wi-Fi/MQTT status LED AND in addition to the power state indication . Led1/Led1i  is ignored (without function).

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT
_Add any other context about the problem here._

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by ascillato** [comment]
Hi,

LED1 is always the status of Relay1.
LED2 is always the status of Relay2.
LEDLINK is the Wi-Fi/MQTT status.

If you only define LED1, that GPIO is going to be LED1+LEDLINK
If you define LED1 and LED2, LED1 is going to be LED1+LEDLINK and LED2 just the status of Relay2.
If you define LEDLINK, the LED1 is going to be just LED1 (the status of Relay1).

Please, in order to achieve what you want, redefine your GPIOs to use:
* LED1 to the gpio that have the led you want to use as relay1 status.
* LEDLINK to the gpio that have the led you want to use as Wi-Fi/MQTT status.

NOTE: The `i` in LED1i means inverted. You can define only one type. You shouldn't assign LED1 to more than one GPIO. Besides that, you shouldn't assign LED1 and LED1i to different GPIOs.

**Reply by schortl** [reply]
Hi,

Regarding your answer my I post you my observations:
(I refer to releases 9.1.0 and 9.5.0 - in release 8.5.1 no Wifi link indication possible whatsoever)


>If you only define LED1, that GPIO is going to be LED1+LEDLINK

true, LED1 shows status of Relay1 + Relay2 + LEDLINK

>If you define LED1 and LED2, LED1 is going to be LED1+LEDLINK and LED2 just the status of Relay2.

**-->different behaviour:**
**LED1 -> no function!!
LED2 -> shows status of Relay1 + Relay2 + LEDLINK**


>If you define LEDLINK, the LED1 is going to be just LED1 (the status of Relay1).

true, ( however, Relay2 status is ignored )

>Please, in order to achieve what you want, redefine your GPIOs to use:

>    LED1 to the gpio that have the led you want to use as relay1 status.
>    LEDLINK to the gpio that have the led you want to use as Wi-Fi/MQTT status

Clear, but wanted behaviour is as it was the case e.g. in release 8.5.1 and as it is described in tasmota.github.io):
If  LED1 and LED2 are defined, LED2 is going to be Relay1 status + Relay2 status and LED1 is Wifi Status:

**LED1 -> LEDLINK
LED2 -> shows status of Relay1 + Relay2**

**or even better as you described above**

**LED1 ->  status Relay1
LED2 -> status of Relay2 + LEDLINK**

>NOTE: The i in LED1i means inverted. You can define only one type. You shouldn't assign LED1 to more than one GPIO. >Besides that, you >shouldn't assign LED1 and LED1i to different GPIOs.

was clear ;)

cheers
shortl

**Reply by ascillato** [reply]
The explanations are for latest release only. Latest Tasmota behaves as stated in Tasmota docs.

**Reply by ascillato** [reply]
In latest Tasmota:

LedX always corresponds to RelayX. If you define LED2 and you don't have a Relay2, that GPIO won't do anything.

If there is no GPIO defined as LedLink, the status of Wi-Fi/MQTT will be in LED1 in addition to the status of relay 1.

If there is a GPIO defined as LEDLINK, then LED1 will only inform relay status.

**Reply by schortl** [reply]
**All my comments relate to the latest release 9.5.0.** 
Checked it again and it behaves:
if no GPIO is defined as LedLink, the status of Wi-Fi/MQTT in addition to the  status of relay 1 is in LED2.
LED1 shows no function and status of relay2 is thus not indicated.
Can you please check? Tried everything (fresh flash etc.)

cheers
shortl

**Reply by barbudor** [reply]
@schortl You seem to be correct
With Led_1 and Led_2 set on my Gosund (same as Nous A1), Led_2 gives both Relay and Wifi status. Nothing seems to drive Led_1

**Reply by ascillato** [reply]
Tested and it is not working. You are right.

It is super weird, I was using all that in recent versions so this Bug is new. Moving to issues.

Thanks for reporting.

**Comment by ascillato** [comment, marked as answer]
Let's move this to a Bug Report.

https://github.com/arendst/Tasmota/issues/13368

Thanks for reporting.

**Reply by barbudor** [reply]
As I said in the Issue, it seems to be an edge case to have 2 leds (and no linkled) defined on a device with a single relay.
I don't think the docs nor the code really cover this case.
